### PR TITLE
[language][move] Add parsing tests [part 2]

### PR DESCRIPTION
**Pull request by bob-wilson**
Add some tests for "let" bindings and expressions.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

I ran these new tests and confirmed that they pass.

### CODE REVIEW

**Comment by bob-wilson** [comment]
I just updated this to put the new tests in the "parser" directory, to match what Todd did for some other tests.

File: language/move-lang/tests/move_check/parser/expr_abort_braces.move [hunk 1]
```diff
@@ -0,0 +1,6 @@
+module M {
+    f(v: u64) {
+        // Braces required for control expression inside "abort" expression
```
Comment (tnowacki) [inline language/move-lang/tests/move_check/parser/expr_abort_braces.move]: I think parens would do too

File: language/move-lang/tests/move_check/parser/expr_if_braces.move [hunk 2]
```diff
@@ -0,0 +1,6 @@
+module M {
+    f(cond: bool) {
+        // Braces required for control expression inside "if" expression
```
Comment (tnowacki) [inline language/move-lang/tests/move_check/parser/expr_if_braces.move]: I think parens would work too

File: language/move-lang/tests/move_check/parser/expr_loop_braces.move [hunk 3]
```diff
@@ -0,0 +1,6 @@
+module M {
+    f(v: u64) {
+        // Braces required for control expression inside "loop" expression
```
Comment (tnowacki) [inline language/move-lang/tests/move_check/parser/expr_loop_braces.move]: ditto parens

File: language/move-lang/tests/move_check/parser/expr_return_braces.move [hunk 4]
```diff
@@ -0,0 +1,6 @@
+module M {
+    f(v: u64): u64 {
+        // Braces required for control expression inside "return" expression
```
Comment (tnowacki) [inline language/move-lang/tests/move_check/parser/expr_return_braces.move]: ditto parens

File: language/move-lang/tests/move_check/parser/expr_return_missing_value.move [hunk 5]
```diff
@@ -0,0 +1,6 @@
+module M {
+    f(v: u64) {
+        // Returns always require a value
```
Comment (tnowacki) [inline language/move-lang/tests/move_check/parser/expr_return_missing_value.move]: We should fix this, probably

Comment (tnowacki) [reply language/move-lang/tests/move_check/parser/expr_return_missing_value.move]: That is, a return without a value should be equivalent to `return ()`

File: language/move-lang/tests/move_check/parser/expr_while_braces.move [hunk 6]
```diff
@@ -0,0 +1,6 @@
+module M {
+    f(v: u64) {
+        // Braces required for assignment inside "while" expression
```
Comment (tnowacki) [inline language/move-lang/tests/move_check/parser/expr_while_braces.move]: ditto parens

File: language/move-lang/tests/move_check/parser/let_binding.move [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+        // FIXME: The grammar currently treats "()" as a single token, but it is
+        // not clear if that has any advantage compared to separate tokens that
+        // allow whitespace in between them.
+        // let ( ) = ( );
+        // let ( ): ( ) = ( );
+        let v1 = 1;
+        let v2: u64 = 2;
+        // FIXME: The grammar does not currently accept a single variable inside
+        // parens, but for consistency, maybe that ought to be allowed.
+        // let (v3) = 3;
```
Comment (tnowacki) [inline language/move-lang/tests/move_check/parser/let_binding.move]: We might want to allow it, yeah. I am undecided

File: language/move-lang/tests/move_check/parser/let_binding.move [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    struct R {
+        f: u64
+    }
+    struct Generic<T> {
+        g: T
+    }
+    f() {
+        let () = ();
+        let (): () = ();
+        // FIXME: The grammar currently treats "()" as a single token, but it is
```
Comment (tnowacki) [inline language/move-lang/tests/move_check/parser/let_binding.move]: The only one is I probably don't want spaces between the type for `unit`, i.e. it should always be `foo(): ()` not `foo(): (   )`

**Comment by bob-wilson** [comment]
I made a follow-up PR to address the comments about using either braces or parens: https://github.com/libra/libra/pull/1965

I also uncommented all the tests that were not supported by the lalrpop parser.

File: language/move-lang/tests/move_check/parser/expr_return_missing_value.move (hunk 5 above)
Comment (bob-wilson) [reply language/move-lang/tests/move_check/parser/expr_return_missing_value.move]: OK. I made a note to myself to look into that.

File: language/move-lang/tests/move_check/parser/let_binding.move (hunk 7 above)
Comment (bob-wilson) [reply language/move-lang/tests/move_check/parser/let_binding.move]: The new parser already allows this, so I went ahead and uncommented the test. If you decide you really don't like it, I'm certainly open to discussing it.

File: language/move-lang/tests/move_check/parser/let_binding.move (hunk 8 above)
Comment (bob-wilson) [reply language/move-lang/tests/move_check/parser/let_binding.move]: I disagree on this one. The new parser treats "(" and ")" as separate tokens. It's just simpler and more consistent that way. We could add a special check of the source locations to detect whitespace between them, but I don't even think that's a good idea. There may be some cases, e.g., code generated from a macro preprocessor where there are types conditionally inserted between the parenthesis, where forbidding whitespace here would be a hassle. I wouldn't ever recommend writing the unit type with whitespace, but it doesn't seem like we need to actively disallow it.

**Pull request closed without merging**